pkg/processor/limiter/processor.go

```diff
@@ -0,0 +1,146 @@
+package limiter
+
+import (
+	"context"
+	"time"
+
+	"github.com/SigNoz/signoz-otel-collector/pkg/entity"
+	"github.com/SigNoz/signoz-otel-collector/pkg/env"
+	"github.com/SigNoz/signoz-otel-collector/pkg/errors"
+	"github.com/SigNoz/signoz-otel-collector/pkg/storage"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type limiterProcessor struct {
+	logger  *zap.Logger
+	storage *storage.Storage
+}
+
+func newLimiterProcessor(logger *zap.Logger) (*limiterProcessor, error) {
+	return &limiterProcessor{
+		logger:  logger,
+		storage: env.G().Storage(),
+	}, nil
+}
+
+func (lp *limiterProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) (ptrace.Traces, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalTraces,
+		entity.NewLimitValue(
+			td.SpanCount(),
+			(&ptrace.ProtoMarshaler{}).TracesSize(td),
+		),
+	)
+
+	if err != nil {
+		return td, err
+	}
+
+	return td, nil
+}
+
+func (lp *limiterProcessor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalMetrics,
+		entity.NewLimitValue(
+			md.MetricCount(),
+			(&pmetric.ProtoMarshaler{}).MetricsSize(md),
+		),
+	)
+
+	if err != nil {
+		return md, err
+	}
+
+	return md, nil
+}
+
+func (lp *limiterProcessor) ConsumeLogs(ctx context.Context, ld plog.Logs) (plog.Logs, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalLogs,
+		entity.NewLimitValue(
+			ld.LogRecordCount(),
+			(&plog.ProtoMarshaler{}).LogsSize(ld),
+		),
+	)
+
+	if err != nil {
+		return ld, err
+	}
+
+	return ld, nil
+}
+
+func (lp *limiterProcessor) consume(ctx context.Context, signal entity.Signal, limitValue entity.LimitValue) error {
+	// Retrieve credential id from ctx
+	keyId, ok := entity.KeyIdFromCtx(ctx)
+	if !ok {
+		return nil
+	}
+
+	// Retrieve the limit from the backing store
+	limit, err := lp.storage.DAO.Limits().SelectByKeyAndSignal(ctx, keyId, signal)
```

Comment:
The limits do not change frequently. It will be fine to fetch the limits once every `fetchInterval`.

Reply:
Users can change limits from the UI. In addition, some users have requested a feature to stop ingestion for a specified period of time for which they can set the limits to 0.

Caching calls to save DB queries can be picked up later as a general theme to improve performance.

Reply:
I know they can change the limits from the UI and can be set to any number. What I am trying to say is that they are not frequent. We pick a sensible fetchInterval and only fetch the details of that interval instead of fetching it for each consume call. Say we choose the `fetchInterval` as 1 minute, now the maximum delay in applying the limits is a maximum of one minute which is fine in all scenarios.

Reply:
This is applicable to all queries, even the query used in the `keyauth` processor. Can pick it up as theme later altogether to improve performance.